File: c_files/dsa/circular_linked_list/circular_linked_list.c
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include "circular_linked_list.h"

Node *create_node(const int data)
{
    Node *node = malloc(sizeof(Node));
    assert(node != NULL);

    node->data = data;
    node->prev = NULL;
    node->next = NULL;

    return node;
}

void cll_init(CircularLinkedList *list)
{
    assert(list);

    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

void cll_prepend(CircularLinkedList *list, const int data)
{
    assert(list);

    Node *new_node = create_node(data);

    if (!list->head) {
        new_node->next = new_node;
        new_node->prev = new_node;

        list->head = new_node;
        list->tail = new_node;

        list->size++;
        return;
    }

    new_node->next = list->head;
    new_node->prev = list->tail;

    list->head->prev = new_node;
    list->tail->next = new_node;

    list->head = new_node;

    list->size++;
}

void cll_append(CircularLinkedList *list, const int data)
{
    assert(list);

    Node *new_node = create_node(data);

    if (!list->head) {
        new_node->next = new_node;
        new_node->prev = new_node;

        list->head = new_node;
        list->tail = new_node;

        list->size++;
        return;
    }

    new_node->next = list->head;
    new_node->prev = list->tail;

    list->head->prev = new_node;
    list->tail->next = new_node;

    list->tail = new_node;

    list->size++;
}

void cll_remove_at(CircularLinkedList *list, const int index)
{
    assert(list && list->head && list->tail);
    assert(list->size > 0);
    assert(index >= 0 && index < list->size);

    Node *current = list->head;
    int   count   = 0;

    while (1) {
        if (count == index) {
            if (list->size == 1) {
                list->head = list->tail = NULL;
                list->size--;
                return;
            }

            if (current->prev) {
                current->prev->next = current->next;
            }
            if (current->next) {
                current->next->prev = current->prev;
            }

            if (current == list->head) {
                list->head = current->next;
            }

            if (current == list->tail) {
                list->tail = current->prev;
            }

            free(current);
            list->size--;
            return;
        }

        current = current->next;
        count++;

        if (current == list->head) {
            break;
        }
    }
}

void cll_remove(CircularLinkedList *list, const int data)
{
    assert(list && list->head && list->tail);
    assert(list->size > 0);

    Node *current = list->head;

    while (1) {
        if (current->data == data) {
            if (list->size == 1) {
                list->head = list->tail = NULL;
                list->size--;
                return;
            }

            if (current->prev) {
                current->prev->next = current->next;
            }
            if (current->next) {
                current->next->prev = current->prev;
            }

            if (current == list->head) {
                list->head = current->next;
            }

            if (current == list->tail) {
                list->tail = current->prev;
            }

            free(current);
            list->size--;
            return;
        }

        current = current->next;

        if (current == list->head) {
            break;
        }
    }

    printf("data not found\n");
}

void cll_reverse(CircularLinkedList *list)
{
    assert(list && list->head && list->tail);
    assert(list->size > 0);

    Node *current = list->head;
    Node *prev    = NULL;
    Node *temp    = NULL;

    while (1) {
        temp = current->next;

        current->prev = temp;
        current->next = prev;

        prev    = current;
        current = temp;

        if (current == list->head) {
            break;
        }
    }

    list->head = current;
    list->tail = current->prev;
}

void cll_print(CircularLinkedList *list)
{
    assert(list && list->head && list->tail);

    if (list->size == 0) {
        printf("empty list\n");
        return;
    }

    printf("Next: tail -> ");

    Node *current = list->head;
    while (1) {
        printf("%d -> ", current->data);

        current = current->next;

        if (current == list->head) {
            break;
        }
    }

    printf("head\n");

    //
    printf("Prev: tail <- ");

    current = list->tail;
    while (1) {
        printf("%d <- ", current->data);

        current = current->prev;

        if (current == list->tail) {
            break;
        }
    }

    printf("head\n");
}

void cll_free(CircularLinkedList *list)
{
    assert(list && list->head && list->tail);
    assert(list->size > 0);

    Node *current = list->head;
    while (1) {
        Node *temp = current;
        current    = current->next;
        free(temp);

        if (current == list->head) {
            break;
        }
    }

    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
}

File: python_files/dsa/circular_linked_list.py
class Node:
    def __init__(self, data: int) -> None:
        self.data: int = data
        self.prev: Node | None = None
        self.next: Node | None = None


class CircularLinkedList:
    def __init__(self) -> None:
        self.size: int = 0
        self.head: Node | None = None
        self.tail: Node | None = None

    def prepend(self, data: int) -> None:
        new_node: Node = Node(data)

        if not self.head:
            new_node.next = new_node
            new_node.prev = new_node

            self.head = new_node
            self.tail = new_node

            self.size += 1
            return

        new_node.next = self.head
        new_node.prev = self.tail

        self.head.prev = new_node
        if self.tail:
            self.tail.next = new_node

        self.head = new_node

        self.size += 1

    def append(self, data: int) -> None:
        new_node: Node = Node(data)

        if not self.head:
            new_node.next = new_node
            new_node.prev = new_node

            self.head = new_node
            self.tail = new_node

            self.size += 1
            return

        new_node.next = self.head
        new_node.prev = self.tail

        self.head.prev = new_node
        if self.tail:
            self.tail.next = new_node

        self.tail = new_node

        self.size += 1

    def remove_at(self, index: int) -> None:
        assert self.head
        assert self.tail
        assert self.size > 0
        assert 0 <= index < self.size

        current: Node | None = self.head
        count: int = 0

        while current:
            if count == index:
                if self.size == 1:
                    self.head = self.tail = None
                    self.size -= 1
                    return

                if current.prev:
                    current.prev.next = current.next
                if current.next:
                    current.next.prev = current.prev

                if current == self.head:
                    self.head = current.next
                if current == self.tail:
                    self.tail = current.prev

                self.size -= 1
                return

            current = current.next
            count += 1

            if current == self.head:
                break

    def remove(self, data: int) -> None:
        assert self.head
        assert self.tail
        assert self.size > 0

        current: Node | None = self.head

        while current:
            if current.data == data:
                if self.size == 1:
                    self.head = self.tail = None
                    self.size -= 1
                    return

                if current.prev:
                    current.prev.next = current.next
                if current.next:
                    current.next.prev = current.prev

                if current == self.head:
                    self.head = current.next
                if current == self.tail:
                    self.tail = current.prev

                self.size -= 1
                return

            current = current.next

            if current == self.head:
                break

        print("data not found")

    def reverse(self) -> None:
        assert self.head
        assert self.tail
        assert self.size > 0

        current: Node | None = self.head
        prev: Node | None = self.tail
        next: Node | None = None

        while current:
            next = current.next

            current.prev = next
            current.next = prev

            prev = current
            current = next

            if current == self.head:
                break

        self.head = current
        self.tail = prev

    def print(self) -> None:
        assert self.head
        assert self.tail
        assert self.size > 0

        print("Next: tail -> ", end="")
        current: Node | None = self.head

        while current:
            print(current.data, end=" -> ")
            current = current.next

            if current == self.head:
                break

        print("head")

        print("Prev: head <- ", end="")
        current = self.tail

        while current:
            print(current.data, end=" <- ")
            current = current.prev

            if current == self.tail:
                break

        print("tail")


def main() -> None:
    arr: CircularLinkedList = CircularLinkedList()
    print()
    print("Append: 1...9")
    for i in range(1, 10):
        arr.append(i)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Prepend: 0")
    arr.prepend(0)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Remove at: 4")
    arr.remove_at(4)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Remove: 8")
    arr.remove(8)
    arr.print()
    print(f"size: {arr.size}")

    print()
    print("Reverse:")
    arr.reverse()
    arr.print()
    print(f"size: {arr.size}")


if __name__ == "__main__":
    main()
